### PR TITLE
[Fix] Move init dist connection into the setup function

### LINKED ISSUE #6318: Default process group is not initialized in setup() function
## 🐛 Bug

Default process group is not initialized in `Datamodule` `setup()` function.

This is a BC breaking with `PL >= 1.2.0`
With, `PL == 1.1.8` this code works.

Reproduce notebook: https://colab.research.google.com/drive/1AHadRi0Bly9OnzrJFv8XmS2T9Y5zklvg?usp=sharing

### Expected behavior
`fit()` should be work.

### Environment
```
* CUDA:
	- GPU:
		- Tesla T4
	- available:         True
	- version:           10.1
* Packages:
	- numpy:             1.19.5
	- pyTorch_debug:     False
	- pyTorch_version:   1.7.1+cu101
	- pytorch-lightning: 1.2.1
	- tqdm:              4.41.1
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- 
	- processor:         x86_64
	- python:            3.7.10
	- version:           #1 SMP Thu Jul 23 08:00:38 PDT 2020
```

### PR DESCRIPTION
**Pull request by SeanNaren**
## What does this PR do?

Fixes #6318 

This fix moves the init ddp connection for DDP into the setup function, and reorders the hook such that setup can now have access to the initialized distributed environment. This is also important for FSDP.

This fix however diverges DDP Spawn from DDP, and should be noted in the docs. As @ananthsub and @awaelchli have brought up, we may need to discuss the responsibility of hooks, primarily because we're seeing some inflexibility in the API when changing call orders of hooks/custom loops. 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by awaelchli** [comment]
The tests in `tests\accererator\test_acceleratror_connector` hang on my server. I'll investigate

**Comment by SeanNaren** [comment]
Few pointers:

- DeepSpeed GPU tests had to be made into special tests, I was getting local errors since torch dist is being initialized and making them independent tests seems to fix them. Not high priority now, but we could wrap them potentially in a fixture such that torch distributed is deleted
- I have to call `model_to_device` twice in the single device plugin; once before and once within pre_dispatch. This is to handle the case if user's define layers within setup which need to be moved to the correct devices. This does mean if the user defines layer in setup, things can break using APEX since weights may not be transferred to half at initialization time. I'm going to investigate further to see if I can remove the `model_to_device` duplicate in the single device plugin, and fix the Apex plugin to handle this case independently

**Comment by awaelchli** [comment]
> DeepSpeed GPU tests had to be made into special tests, I was getting local errors since torch dist is being initialized and making them independent tests seems to fix them. Not high priority now, but we could wrap them potentially in a fixture such that torch distributed is deleted

We can't terminate the ddp connection after a trainer.fit call because we don't know if  the user will call .fit or .test again right after it, and we want to enable that. The connection staying open could be a problem in pytest the way these tests are launched and stopped. A fixture could help us in the tests ensuring that no connections are open before running a new test, while still enabling the use case described above outside of the test environment. 
One thing I would test is this: Take one of these deepspeed tests and copy the contents to a regular script and execute it. Does leave any connections open after it ends?

**Comment by awaelchli** [comment]
> I have to call model_to_device twice in the single device plugin; once before and once within pre_dispatch. This is to handle the case if user's define layers within setup which need to be moved to the correct devices. This does mean if the user defines layer in setup, things can break using APEX since weights may not be transferred to half at initialization time. I'm going to investigate further to see if I can remove the model_to_device duplicate in the single device plugin, and fix the Apex plugin to handle this case independently

Maybe it helps to go back to before the accelerator refactor and look again what the order of calls were. The commit before acc refactor is 309ce7a96664316b0f42147d6a849001d841559a
In the DDPAccelerator for example we can see the model is moved to the device after the setup hook is called.

File: pytorch_lightning/accelerators/accelerator.py [hunk 1]
```diff
@@ -73,9 +85,9 @@ def setup(self, trainer: 'Trainer', model: LightningModule) -> None:
             trainer: the trainer instance to connect to
             model: the model to train
         """
-        self.connect_training_type_plugin(self.training_type_plugin, model)
+        self.setup_training_type_plugin(self.training_type_plugin, model)
         self.setup_optimizers(trainer)
-        self.connect_precision_plugin(self.precision_plugin)
+        self.setup_precision_plugin(self.precision_plugin)
```
Comment (SeanNaren) [inline pytorch_lightning/accelerators/accelerator.py:90]: If anyone extended and made their own accelerator, this will be a breaking change so might need to handle a deprecation path here

File: pytorch_lightning/plugins/training_type/ddp.py [hunk 2]
```diff
@@ -213,37 +241,6 @@ def init_ddp_connection(self, global_rank: int, world_size: int) -> None:
...
-
-        # set up server using proc 0's ip address
-        # try to init for 20 times at max in case ports are taken
-        # where to store ip_table
-        self.init_ddp_connection(self.global_rank, self.world_size)
-
-        # TODO: we moved it to the trainer.fit after calling pre_dispatch
-        #   ... need to double check that it is the correct place
-        # self.trainer.call_setup_hook(self.model)
-
```
Comment (awaelchli) [inline pytorch_lightning/plugins/training_type/ddp.py:235]: yeah my silly todo....
"need to double check that it is the correct place"

Thanks for double checking @SeanNaren 😄

File: tests/special_tests.sh [hunk 3]
```diff
@@ -17,6 +17,11 @@ export PL_RUNNING_SPECIAL_TESTS=1
 DEFAULTS="-m coverage run --source pytorch_lightning -a -m pytest --verbose --capture=no"
 python ${DEFAULTS} tests/trainer/optimization/test_manual_optimization.py::test_step_with_optimizer_closure_with_different_frequencies_ddp
 python ${DEFAULTS} tests/models/test_sync_batchnorm.py::test_sync_batchnorm_ddp
+python ${DEFAULTS} tests/plugins/test_deepspeed_plugin.py::test_warn_deepspeed_override_backward
+python ${DEFAULTS} tests/plugins/test_deepspeed_plugin.py::test_deepspeed_run_configure_optimizers
+python ${DEFAULTS} tests/plugins/test_deepspeed_plugin.py::test_deepspeed_config
+python ${DEFAULTS} tests/plugins/test_deepspeed_plugin.py::test_deepspeed_custom_precision_params
+python ${DEFAULTS} tests/plugins/test_deepspeed_plugin.py::test_deepspeed_assert_config_zero_offload_disabled
```
Comment (awaelchli) [inline tests/special_tests.sh:24]: above you set 6 tests with special=True but here are only 5

File: tests/accelerators/test_accelerator_connector.py [hunk 4]
```diff
@@ -224,7 +228,8 @@ def on_fit_start(self, trainer, pl_module):
     "NODE_RANK": "0",
 })
 @mock.patch('torch.cuda.device_count', return_value=0)
-def test_accelerator_choice_ddp_cpu_te(device_count_mock):
+@mock.patch('pytorch_lightning.plugins.DDPPlugin.setup_distributed', autospec=True)
```
Comment (awaelchli) [inline tests/accelerators/test_accelerator_connector.py:231]: I didn't know that this works. I always thought that one has to give the full path :)

**Review by awaelchli — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM ! Great job !

File: pytorch_lightning/accelerators/accelerator.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/accelerators/accelerator.py:90]: Do you need to rename them ?

Comment (SeanNaren) [reply pytorch_lightning/accelerators/accelerator.py:90]: I don't i guess? It's just a bit concerning because if I don't rename them, `connect_training_type_plugin` will be calling `plugin.setup`, and `connect` will be calling `training_type_plugin.connect`. Just confusing function names. I think if it becomes an issue we can make this BW compatible however in most cases it seems users should be defining plugins, not accelerators.

**Review by ananthsub — APPROVED** [review]
🚀

File: pytorch_lightning/plugins/training_type/ddp.py [hunk 5]
```diff
@@ -161,6 +161,34 @@ def _call_children_scripts(self):
...
+        # try to init for 20 times at max in case ports are taken
+        # where to store ip_table
+        self.init_ddp_connection(self.global_rank, self.world_size)
+
+        # on world_size=0 let everyone know training is starting
+        if self.is_global_zero and not torch.distributed.is_initialized():
+            log.info("-" * 100)
+            log.info(f"distributed_backend={self.distributed_backend}")
+            log.info(f"All DDP processes registered. Starting ddp with {self.world_size} processes")
+            log.info("-" * 100)
```
Comment (Borda) [inline pytorch_lightning/plugins/training_type/ddp.py:186]: btw, shall we have this as a single message intend for 4 separate?

File: tests/plugins/test_deepspeed_plugin.py [hunk 6]
```diff
@@ -180,7 +180,7 @@ def test_deepspeed_defaults(tmpdir):
     assert isinstance(plugin.config["zero_optimization"], dict)
 
 
-@RunIf(deepspeed=True)
+@RunIf(min_gpus=1, deepspeed=True, special=True)
```
Comment (Borda) [inline tests/plugins/test_deepspeed_plugin.py:183]: this is so cool :D

**Comment by carmocca** [comment]
Setting milestone as 1.3 as it requires separating the fix from the API change to get into `1.2.x`.
